simple/multi_mr.c

```diff
@@ -0,0 +1,496 @@
+	if (ret) {
+		FT_PRINTERR("fi_accept", ret);
+		goto err;
+	}
+
+	rd = fi_eq_sread(eq, &event, &entry, sizeof(entry), -1, 0);
+	if (rd != sizeof(entry)) {
+		FT_PROCESS_EQ_ERR(rd, eq, "fi_eq_sread", "accept");
+		ret = (int) rd;
+		goto err;
+	}
+
+	if (event != FI_CONNECTED || entry.fid != &ep->fid) {
+		fprintf(stderr, "Unexpected CM event %d fid %p (ep %p)\n",
+			event, entry.fid, ep);
+		ret = -FI_EOTHER;
+		goto err;
+	}
+
+	return 0;
+
+err:
+	fi_reject(pep, fi->handle, NULL, 0);
+	return ret;
+}
+
+static int client_connect(void)
+{
+	struct fi_eq_cm_entry entry;
+	uint32_t event;
+	ssize_t rd;
+	int ret;
+
+	ret = ft_getinfo(hints, &fi);
+	if (ret)
+		return ret;
+
+	ret = ft_open_fabric_res();
+	if (ret)
+		return ret;
+
+	/* Prevent memory registration by ft_alloc_active_res() -> ft_alloc_msgs() */
+	ft_skip_mr = 1;
+
+	ret = ft_alloc_active_res(fi);
+	if (ret)
+		return ret;
+
+	ret = alloc_multi_mr_res();
+	if (ret)
+		return ret;
+
+	ret = ft_init_ep();
+	if (ret)
+		return ret;
+
+	ret = fi_connect(ep, fi->dest_addr, NULL, 0);
+	if (ret) {
+		FT_PRINTERR("fi_connect", ret);
+		return ret;
+	}
+
+	rd = fi_eq_sread(eq, &event, &entry, sizeof(entry), -1, 0);
+	if (rd != sizeof(entry)) {
+		FT_PROCESS_EQ_ERR(rd, eq, "fi_eq_sread", "connect");
+		ret = (int) rd;
+		return ret;
+	}
+
+	if (event != FI_CONNECTED || entry.fid != &ep->fid) {
+		fprintf(stderr, "Unexpected CM event %d fid %p (ep %p)\n",
+			event, entry.fid, ep);
+		ret = -FI_EOTHER;
+		return ret;
+	}
+
+	return 0;
+}
+
+static int init_fabric_cm(void)
+{
+	int ret;
+	if (!opts.dst_addr) {
+		ret = ft_start_server();
+		if (ret)
+			return ret;
+	}
+
+	ret = opts.dst_addr ? client_connect() : server_connect();
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static int run_test(void)
+{
+	int ret = 0;
+
+	if (hints->ep_attr->type == FI_EP_MSG) {
+		ret = init_fabric_cm();
+		if (ret)
+			return ret;
+	} else {
+		ret = init_fabric();
+		if (ret)
+			return ret;
+	}
+
+	ret = mr_key_test();
+
+	ft_sync();
+	ft_finalize();
+
+	return ret;
+}
+
+int main(int argc, char **argv)
+{
+	int op;
+	int ret = 0;
+
+	opts = INIT_OPTS;
+	opts.transfer_size = 4096;
+	g_test_state.mr_count = 1;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return EXIT_FAILURE;
+
+	while ((op = getopt(argc, argv, "c:h" ADDR_OPTS INFO_OPTS)) != -1) {
+		switch (op) {
+		default:
+			ft_parse_addr_opts(op, optarg, &opts);
+			ft_parseinfo(op, optarg, hints);
+			break;
+		case 'c':
+			g_test_state.mr_count = atoi(optarg);
+			break;
+		case '?':
+		case 'h':
+			ft_usage(argv[0], "Ping-pong multi memory region test");
+			FT_PRINT_OPTS_USAGE("-c <int>", "power of 2 number of memory regions to create and test");
+			return EXIT_FAILURE;
+		}
+	}
+
+	if (optind < argc)
+		opts.dst_addr = argv[optind];
+
+	hints->caps = FI_RMA | FI_RMA_EVENT | FI_MSG;
+	hints->mode = FI_CONTEXT | FI_LOCAL_MR;
+	hints->tx_attr->op_flags = FI_DELIVERY_COMPLETE;
+	hints->domain_attr->mr_key_size = sizeof(uint64_t);
```

Comment:
We should have the test adjust to the provider's key size.  Most HW only supports 32-bit keys.